common/src/block_interval.rs

```diff
@@ -0,0 +1,215 @@
+        } else if let Ok(range) = parsing::parse_range_exclusive(s) {
+            Ok(BlockInterval::Range(range))
+        }
+        // Now we look for the follow from range
+        else if s.contains("..") {
+            let mut split = s.trim().split("..").filter(|s| *s != "=" && !s.is_empty());
+
+            // Any other character after `..` or `..=` is invalid
+            if split.clone().count() > 1 {
+                return Err(BlockIntervalError::InvalidFollowFromRange(s.to_string()));
+            }
+            let num = split
+                .next()
+                .map(|num| {
+                    num.parse::<u64>()
+                        .map_err(|_| BlockIntervalError::InvalidBlockNumber(s.to_string()))
+                })
+                .ok_or(BlockIntervalError::InvalidBlockIntervalRange(s.to_string()))??;
+            return Ok(BlockInterval::FollowFrom(num));
+        }
+        // Only single block number is left to try to parse
+        else {
+            let num = s
+                .trim()
+                .parse()
+                .map_err(|_| BlockIntervalError::InvalidBlockIntervalRange(s.to_string()))?;
+            return Ok(BlockInterval::Single(num));
+        }
+    }
+
+    /// Convert the block interval into a stream of block numbers.
+    pub fn into_bounded_stream(self) -> Result<impl Stream<Item = u64>, BlockIntervalError> {
+        match self {
+            BlockInterval::Single(num) => Ok(tokio_stream::iter(num..num)),
+            BlockInterval::Range(range) => Ok(tokio_stream::iter(range)),
+            _ => Err(BlockIntervalError::IntoBoundedStreamError),
+        }
+    }
+
+    /// Convert the block interval into an unbounded stream of block numbers.
+    /// Query the blockchain node for the latest block number
+    pub async fn into_unbounded_stream<ProviderT, TransportT>(
+        self,
+        provider: ProviderT,
+    ) -> Result<impl Stream<Item = Result<u64, anyhow::Error>>, BlockIntervalError>
+    where
+        ProviderT: Provider<TransportT>,
+        TransportT: Transport + Clone,
+    {
+        match self {
+            BlockInterval::FollowFrom(start) => Ok(try_stream! {
+                let mut current = start;
+                while let Ok(last_block_number) = provider.get_block_number()
+                    .await
+                    .map_err(|e: alloy::transports::RpcError<_>| {
+                        BlockIntervalError::LatestBlockNumberError(e.to_string())
+                    }) {
+                        if current < last_block_number {
+                            current += current;
+                            yield current;
+                        } else {
+                           info!("Waiting for the new blocks to be mined, requested block number: {current}, \
+                           latest block number: {last_block_number}");
+                           tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;
```

Comment:
This feels like it could be nice to be configurable.

Reply:
@muursh It is [here](https://github.com/0xPolygonZero/zero-bin/pull/90/commits/9518616fdd2df0c269c6cf33439b03f32c833251)